git-version-macro/src/describe_submodules.rs

```diff
@@ -0,0 +1,166 @@
+use crate::git_dependencies;
+use crate::utils::run_git;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use std::ffi::OsStr;
+use std::path::Path;
+use std::process::Command;
+use syn::{
+	bracketed,
+	parse::{Parse, ParseStream},
+	punctuated::Punctuated,
+	token::{Comma, Eq},
+	Expr, Ident, LitStr,
+};
+
+macro_rules! error {
+	($($args:tt)*) => {
+		syn::Error::new(Span::call_site(), format!($($args)*))
+	};
+}
+
+#[derive(Default)]
+pub(crate) struct GitModArgs {
+	args: Option<Punctuated<LitStr, Comma>>,
+	prefix: Option<LitStr>,
+	suffix: Option<LitStr>,
+	fallback: Option<Expr>,
+}
+
+impl Parse for GitModArgs {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		let mut result = GitModArgs::default();
+		loop {
+			if input.is_empty() {
+				break;
+			}
+			let ident: Ident = input.parse()?;
+			let _: Eq = input.parse()?;
+			let check_dup = |dup: bool| {
+				if dup {
+					Err(error!("`{} = ` can only appear once", ident))
+				} else {
+					Ok(())
+				}
+			};
+			match ident.to_string().as_str() {
+				"args" => {
+					check_dup(result.args.is_some())?;
+					let content;
+					bracketed!(content in input);
+					result.args = Some(Punctuated::parse_terminated(&content)?);
+				}
+				"prefix" => {
+					check_dup(result.prefix.is_some())?;
+					result.prefix = Some(input.parse()?);
+				}
+				"suffix" => {
+					check_dup(result.suffix.is_some())?;
+					result.suffix = Some(input.parse()?);
+				}
+				"fallback" => {
+					check_dup(result.fallback.is_some())?;
+					result.fallback = Some(input.parse()?);
+				}
+				x => Err(error!("Unexpected argument name `{}`", x))?,
+			}
+			if input.is_empty() {
+				break;
+			}
+			let _: Comma = input.parse()?;
+		}
+		Ok(result)
+	}
+}
+
+pub(crate) fn git_module_versions_impl(args: GitModArgs) -> syn::Result<TokenStream2> {
+	let mut modules = match get_modules() {
+		Ok(x) => x,
+		Err(err) => return Err(error!("{}", err)),
+	};
+	modules.retain(|path| path != "");
+	println!("modules:{:?}", modules);
+	let mut describe_paths: Vec<(String, String)> = vec![];
+
+	for path in modules.into_iter() {
+		let path_obj = Path::new(&path);
+		let path_obj = canonicalize_path(path_obj)?;
+		describe_paths.push((path, path_obj));
+	}
+
+	let git_describe_args = args.args.map_or_else(
+		|| vec!["--always".to_string(), "--dirty=-modified".to_string()],
+		|list| list.iter().map(|x| x.value()).collect(),
+	);
+
+	let prefix = match args.prefix {
+		Some(x) => x.value(),
+		_ => "".to_string(),
+	};
+	let suffix = match args.suffix {
+		Some(x) => x.value(),
+		_ => "".to_string(),
+	};
+
+	match describe_modules(describe_paths, &git_describe_args, prefix, suffix) {
+		Ok(result) => {
+			let dependencies = git_dependencies()?;
+			let (paths, versions) = result;
+
+			Ok(quote!({
+				#dependencies;
+
+				[#((#paths, #versions)),*]
+
+			}))
+		}
+		Err(_) if args.fallback.is_some() => Ok(quote!([("fallback", args.fallback)])),
```

Comment:
I'm not sure this makes sense here. For `git_version!()`, the fallback means: if "git describe fails, give this as version instead".

I think the fallback should have the same meaning for submodules: each submodule entry should apply the fallback independently if `git describe` fails.

At the very least, it will mean that the submodule paths are always real paths. With the current interpretation of the fallback argument, you can't distinguish between the fallback being used or a submodule at path "fallback".

Reply:
I thought it felt clunky. I will implement this tomorrow.